agent/consul/state/acl.go

```diff
@@ -25,10 +62,186 @@ func (s *Restore) ACL(acl *structs.ACL) error {
 	if err := indexUpdateMaxTxn(s.tx, acl.ModifyIndex, "acls"); err != nil {
 		return fmt.Errorf("failed updating index: %s", err)
 	}
+	return nil
+}
+
+// ACLBootstrap is used to pull the ACL bootstrap info from the snapshot. This
+// might return nil, in which case nothing should be saved to the snapshot.
+func (s *Snapshot) ACLBootstrap() (*structs.ACLBootstrap, error) {
+	existing, err := s.tx.First("acls-bootstrap", "id")
+	if err != nil {
+		return nil, fmt.Errorf("failed acl bootstrap lookup: %s", err)
+	}
+	if existing != nil {
+		return existing.(*structs.ACLBootstrap), nil
+	}
+	return nil, nil
+}
+
+// ACLBootstrap is used to restore the ACL bootstrap info from the snapshot.
+func (s *Restore) ACLBootstrap(bs *structs.ACLBootstrap) error {
+	if err := s.tx.Insert("acls-bootstrap", bs); err != nil {
+		return fmt.Errorf("failed updating acl bootstrap: %v", err)
+	}
+	return nil
+}
+
+// ACLBootstrapInit is used to perform a scan for existing tokens which will
+// decide whether bootstrapping is allowed for a cluster. This is initiated by
+// the leader when it steps up, if necessary. This is because the state store
+// snapshots would become incompatible with older agents if we added this on
+// the fly, so we rely on the leader to determine a safe time to add this so
+// we can start tracking whether bootstrap is enabled. This will return an
+// error if bootstrap is already initialized.
+//
+// This returns a boolean indicating if ACL boostrapping is enabled.
+func (s *Store) ACLBootstrapInit(idx uint64) (bool, error) {
+	tx := s.db.Txn(true)
+	defer tx.Abort()
+
+	// Don't allow this to happen more than once.
+	existing, err := tx.First("acls-bootstrap", "id")
+	if err != nil {
+		return false, fmt.Errorf("failed acl bootstrap lookup: %s", err)
+	}
+	if existing != nil {
+		return false, fmt.Errorf("acl bootstrap init already done")
+	}
+
+	// See if there are any management tokens, which means we shouldn't
+	// allow bootstrapping.
+	foundMgmt, err := s.aclHasManagementTokensTxn(tx)
+	if err != nil {
+		return false, fmt.Errorf("failed checking for management tokens: %v", err)
+	}
+	allowBootstrap := !foundMgmt
+
+	// Create a new bootstrap record.
+	bs := structs.ACLBootstrap{
+		AllowBootstrap: allowBootstrap,
+		RaftIndex: structs.RaftIndex{
+			CreateIndex: idx,
+			ModifyIndex: idx,
+		},
+	}
+	if err := tx.Insert("acls-bootstrap", &bs); err != nil {
+		return false, fmt.Errorf("failed creating acl bootstrap: %v", err)
+	}
+
```

Comment:
There's a different set of pre-checks that are done in each case, so we have these two different methods.